### ISSUE TITLE
Memory leak when using Consul & vault. Linux Memory consumed 

### ISSUE THREAD

**Issue by NaseemMa** [hashicorp/consul#2458]
### `consul version` for both Client and Server
Client: 0.6.4
Server: 0.6.4

Server:
```
[ Consul server with Vault 0.5.3]
```

### Operating system and Environment details
Linux, Redhat 6.5
### Description of the Issue (and unexpected/desired result)
when started Consul & Vault together system memory will be consumed, ~10 MB per minute, we can see that by using #free -m the Mem used increased by the time.
### Reproduction steps
install consul and vault and strart them
monitor the system memory
### Log Fragments or Link to [gist](https://gist.github.com/)

Include appropriate Client or Server log fragments.  If the log is longer
than a few dozen lines, please include the URL to the
[gist](https://gist.github.com/).

TIP: Use `-log-level=TRACE` on the client and server to capture the maximum log detail.

**Comment by slackpad** [comment]
Hi @NaseemMa thanks for opening an issue. The best way to see where the memory is going is to set the enable_debug option on one of your servers and then look at http://localhost:8500/debug/pprof/heap?debug=1 on that server to see the heap profile. Would you be able to post a few samples of that endpoint's output over time as a gist?

**Comment by NaseemMa** [comment]
Thanks,
I already share with you the memstate 
![image](https://cloud.githubusercontent.com/assets/23191465/19892174/4f7317d8-a04c-11e6-88d2-1eacc2ce0cb8.png)

**Comment by slackpad** [comment]
The heap profile is super useful for figuring out where the memory is being used.

**Comment by NaseemMa** [comment]
[heap_1.txt](https://github.com/hashicorp/consul/files/566056/heap_1.txt)
[heap_2.txt](https://github.com/hashicorp/consul/files/566058/heap_2.txt)

**Comment by NaseemMa** [comment]
Hi,
this memory issue happen when using redhat -6.5 version.
on other servers which have CentOS 7 there is no Memory leak.

did you know if there is a BUG on RHL 6.5?

**Comment by slackpad** [comment]
Hi @NaseemMa I took a look through the heap dumps and it looks like there's a lot of KV set operations as well as ListKeys operations. What kind of workload is this cluster doing? Is there some process that could be causing a lot of rapid KV churn with a corresponding blocking query for listing keys? Is it just Vault using Consul?

**Comment by NaseemMa** [comment]
Hi James Phillips,
on that time that i took the heap dumps, there was no process that ran on the cluster, only vault & consul was up.

**Comment by slackpad** [comment]
Hmm - we don't know of any RHL bugs like this. It would be interesting to see if there was something doing a bunch of operations against Vault at the time.

**Comment by sabag** [comment]
Hi,
I am working with NaseemMa, and I would like to contribute more information on the issue.
We ran only Consul and vault processes and no other traffic was recorded (tcpdump). the network dump only showed the same request going from vault to consul something like /renew/session every 7-8 seconds. With that we saw memory going down.
Are there any things that you would like us to provide in order to get to the bottom of this issue. its a very critical issue for us.
Thank you much in advance.
Daniel

**Comment by slackpad** [comment]
Hi @sabag a 7-8 second period for the session renew makes sense for Vault maintaining its leader lock (that has a 15 second TTL and the periodic fires at 1/2 that interval). Do you know the time interval between the two heap dumps you provided? I didn't see a dramatic difference in them. It might be useful to take them several hours apart. I'd also be curious if the memory usage didn't level off after some time - do you have any plots of memory usage over several hours?

I couldn't find any RHEL-related memory leak issues, but it might also be worth building Consul from master with Go 1.7.3 and see if that makes a difference. The next version of Consul will be built with that, and it's several versions newer than what we built Consul 0.6.4 with.

**Comment by NaseemMa** [comment]
I`m Installing Consul and vault on new ENV (same redhat).
is there possibility to install only 1 server or we need minimum 3?

**Comment by slackpad** [comment]
1 is fine for testing but you'd want 3 or 5 for a real production setup. 3 can handle 1 sever failure and 5 can handle 2.

**Comment by NaseemMa** [comment]
Hi,
I installed 1 NODE (server)
Consul & Vault are up, I see that the used memory is growing.
attached 2 heap file. will attach other later.
thanks for you response and help
[heap_9_NOV_17_59_1NODE.txt](https://github.com/hashicorp/consul/files/581262/heap_9_NOV_17_59_1NODE.txt)
[heap_9_NOV_16_56_1NODE.txt](https://github.com/hashicorp/consul/files/581263/heap_9_NOV_16_56_1NODE.txt)

**Comment by sabag** [comment]
Hi all,

We finally found the root cause for the memory leak. The great news is that there is nothing wrong with the Consul/Vault code. The root cause for the memory consumption was curl shell command with https request will leak memory until all memory is consumed.

doing a simple bash loop with curl will show memory dropping.

So, thank you everyone and sorry for the trouble we caused in this issue.
Daniel
